### PR TITLE
use skopeo+tar to get image labels = no root

### LINKED ISSUE #150: colin without the need of dockerd
We would like to explore a solution, where colin would use `skopeo` to perform some checks on labels. This way, colin would be easily able to run in an openshift pod.

### PR DESCRIPTION
**Pull request by TomasTomecek**
...is required so it can run in openshift now

    check images without docker

    use skopeo+tar to get image labels = no root

    The code is hideous. The plan is to give this a shot and if it works,
    move the low level code to conu.

    Also disabling dynamic checks: colin/issues/154

    They are not being utilized yet anyway.

Fixes #150

TODO:
 * [x] filesystem access via `ATOMIC_OSTREE_REPO=$PWD/c atomic pull --storage ostree registry.access.redhat.com/rhscl/mariadb && atomic mount`
 * [x] integration test that this can be run via unpriv user

### CODE REVIEW

**Comment by TomasTomecek** [comment]
Hm, tests pass for me locally:
```
59 passed in 227.54 seconds
```

**Comment by TomasTomecek** [comment]
Tests are failing on selinux:
```
  File "/usr/lib/python3.6/site-packages/Atomic/syscontainers.py", line 2587, in _ensure_storage_for_image
    self._create_rootfs(rootfs)
  File "/usr/lib/python3.6/site-packages/Atomic/syscontainers.py", line 241, in _create_rootfs
    selinux.setfscreatecon_raw(None)
OSError: [Errno 22] Invalid argument
```

File: colin/utils/cont.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+            j = json.load(fd)
+
+        # figure out name of the metadata file
+        metadata_file_name = j[0]["Config"]
+
+        # then extract the metadata
+        extract_file_from_tarball(archive_file_name, metadata_file_name, self.tmpdir)
+        metadata_file_path = os.path.join(self.tmpdir, metadata_file_name)
+        with open(metadata_file_path) as fd:
+            self.metadata = json.load(fd)
```
Comment (jpopelka) [inline colin/utils/cont.py]: Just refactoring nitpick:
these 4 lines could be in a method/function since they're twice here

Comment (TomasTomecek) [reply colin/utils/cont.py]: will do

**Comment by TomasTomecek** [comment]
/o\ I forgot to rebase. Trying again locally.

File: colin/utils/cont.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+logger = logging.getLogger(__name__)
+
+
+def extract_file_from_tarball(tarball_path, file_path, wd):
+    """
+    Extract selected file (file_path) from tarball (tarball_path) within a selected
+    directory (wd).
+    """
+    tar_cmd = ["tar", "-xf", tarball_path, file_path]
+    out = subprocess.check_output(tar_cmd, cwd=wd)
```
Comment (phracek) [inline colin/utils/cont.py]: What about to log also error in case `image` is corrupted/damaged?

File: colin/utils/cont.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+        self._labels = None
+
+        self._pull_image()
+
+
+    @property
+    def tmpdir(self):
+        """ Temporary directory holding all the runtime data. """
+        if self._tmpdir is None:
+            self._tmpdir = mkdtemp(prefix="colin-", dir="/var/tmp")
```
Comment (phracek) [inline colin/utils/cont.py]: Would it be possible to generate tmp dir by NamesTemporaryDirectory?

File: colin/utils/cont.py [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+        metadata_file_path = os.path.join(self.tmpdir, metadata_file_name)
+        with open(metadata_file_path) as fd:
+            self.metadata = json.load(fd)
+
+    def _checkout(self):
+        """ check out the image filesystem on self.mount_point """
+        e = os.environ.copy()
+        e["ATOMIC_OSTREE_REPO"] = self.ostree_path
+        # self.mount_point has to be created by us
+        out = subprocess.check_output(
```
Comment (phracek) [inline colin/utils/cont.py]: debug also error, in case of failure.

File: colin/utils/cont.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+            ["atomic", "mount", "--storage", "ostree", self.image_name, self.mount_point],
+            env=e)
+        logger.debug("output of atomic command:")
+        logger.debug(out)
+
+    def clean_up(self):
+        e = os.environ.copy()
+        e["ATOMIC_OSTREE_REPO"] = self.ostree_path
+        out = subprocess.check_output(["atomic", "unmount", self.mount_point], env=e)
```
Comment (phracek) [inline colin/utils/cont.py]: The same here.

File: colin/core/checks/filesystem.py [hunk 6]
```diff
@@ -37,17 +36,66 @@ def __init__(self, message, description, reference_url, tags, files, all_must_be
...
-    def check(self, target):
+    def _handle_image(self, target):
+        passed = self.all_must_be_present
+
+        logs = []
+        for f in self.files:
+            try:
+                f_present = target.instance.file_is_present(f)
+                logs.append("File '{}' is {}present."
+                            .format(f, "" if f_present else "not "))
```
[the remark this answers is not in the document]
Comment (TomasTomecek) [reply colin/core/checks/filesystem.py]: because method `file_is_present` can raise an exc

File: colin/utils/cont.py (hunk 3 above)
Comment (TomasTomecek) [reply colin/utils/cont.py]: I moved it to cont.py

File: colin/utils/cont.py (hunk 4 above)
Comment (TomasTomecek) [reply colin/utils/cont.py]: +1

File: colin/utils/cont.py (hunk 5 above)
Comment (TomasTomecek) [reply colin/utils/cont.py]: +1

File: colin/utils/cont.py (hunk 2 above)
Comment (TomasTomecek) [reply colin/utils/cont.py]: Good point, I replaced all subprocess calls with a single function which logs everything.

**Comment by dhodovsk** [comment]
I still get an error when running checks on Dockerfiles: :(

```
[root@d242088793b0 colin]# colin check Dockerfile.tests 
13:00:59.275 cont.py           ERROR  Failed to pull selected container image. Does it exist?
13:00:59.276 colin.py          ERROR  An error occurred: CalledProcessError(1, ['atomic', 'pull', '--storage', 'ostree', 'docker:Dockerfile.tests:latest'])
Error: Command '['atomic', 'pull', '--storage', 'ostree', 'docker:Dockerfile.tests:latest']' returned non-zero exit status 1.
```

**Comment by TomasTomecek** [comment]
@dhodovsk that's expected actually: there is a new option now: you need to run:
```
$ colin check --target-type dockerfile ./Dockerfile
```
the option defaults to image, hence the error above

**Comment by dhodovsk** [comment]
Could the error message contain information about possible missing target?

**Comment by TomasTomecek** [comment]
Is this better?
```
$ python3 -m colin.cli.colin check --target-type=image Dockerfile.tests
10:44:08.709 cont.py           ERROR  Failed to pull selected container image. Does it exist?
10:44:08.710 target.py         ERROR  Please make sure that you picked the correct target type: --target-type CLI option.
10:44:08.710 colin.py          ERROR  An error occurred: CalledProcessError(1, ['atomic', 'pull', '--storage', 'ostree', 'docker:Dockerfile.tests:latest'])
Error: Command '['atomic', 'pull', '--storage', 'ostree', 'docker:Dockerfile.tests:latest']' returned non-zero exit status 1.

$ python3 -m colin.cli.colin check --target-type=image asdqwemnb
10:44:32.881 cont.py           ERROR  Failed to pull selected container image. Does it exist?
10:44:32.882 target.py         ERROR  Please make sure that you picked the correct target type: --target-type CLI option.
10:44:32.882 colin.py          ERROR  An error occurred: CalledProcessError(1, ['atomic', 'pull', '--storage', 'ostree', 'docker:asdqwemnb:latest'])
Error: Command '['atomic', 'pull', '--storage', 'ostree', 'docker:asdqwemnb:latest']' returned non-zero exit status 1.

$ python3 -m colin.cli.colin check --target-type=dockerfile fedora:27
10:44:39.239 target.py         ERROR  Please make sure that you picked the correct target type: --target-type CLI option.
10:44:39.240 colin.py          ERROR  An error occurred: FileNotFoundError(2, 'No such file or directory')
Error: [Errno 2] No such file or directory: 'fedora:27'
```

**Comment by lslebodn** [comment]
It would be good to also support insecure registries.

I know you filed a bug to skopeo  https://github.com/projectatomic/skopeo/issues/530
but meanwhile we can use workaound

```
sh-4.4$ diff -u /usr/local/lib/python3.6/site-packages/colin/utils/cont.py colin/utils/cont.py 
--- /usr/local/lib/python3.6/site-packages/colin/utils/cont.py  2018-07-24 14:14:03.000000000 +0000
+++ colin/utils/cont.py 2018-07-30 18:48:05.046559847 +0000
@@ -105,7 +105,7 @@
 
         if self.pull:
             skopeo_source = "docker://" + image_name.name
-            atomic_source = image_name.name
+            atomic_source = "http:" + image_name.name
         else:
             skopeo_source = "docker-daemon:" + image_name.name
             atomic_source = "docker:" + image_name.name
@@ -117,7 +117,7 @@
         archive_file_name = "archive.tar"
         archive_path = os.path.join(self.tmpdir, archive_file_name)
         skopeo_target = "docker-archive:" + archive_path
-        skopeo_cmd = ["skopeo", "copy", skopeo_source, skopeo_target]
+        skopeo_cmd = ["skopeo", "copy", "--src-tls-verify=false", skopeo_source, skopeo_target]
         run_and_log(skopeo_cmd, None,
                     "Failed to create tarball with layers from the selected image")
 
@@ -140,7 +140,7 @@
 
     def _checkout(self):
         """ check out the image filesystem on self.mount_point """
-        cmd = ["atomic", "mount", "--storage", "ostree", self.image_name, self.mount_point]
+        cmd = ["atomic", "mount", "--storage", "ostree", "http:" + self.image_name, self.mount_point]
         # self.mount_point has to be created by us
         run_and_log(cmd, self.ostree_path, "Failed to mount selected image as an ostree repo.")
```

Just use some command line option for insecure registry instead of hardcoded values :-)

**Comment by TomasTomecek** [comment]
Tests are passing locally:
```
58 passed in 196.33 seconds
```

File: colin/core/target.py [hunk 7]
```diff
@@ -187,79 +208,3 @@ class TargetType(enum.Enum):
     DOCKERFILE = 0
     CONTAINER = 1
     IMAGE = 2
-
```
[the remark this answers is not in the document]
Comment (jpopelka) [reply colin/core/target.py]: It'll forever be in our hearts ... I mean ... git history.

**Comment by TomasTomecek** [comment]
open the champagne!

**Pull request merged**